src/synapse.ts

```diff
@@ -1,79 +1,410 @@
 /**
- * Mock implementation of the main Synapse class
+ * Main Synapse class for interacting with Filecoin storage and other on-chain services
  */
 
-import type { Synapse as ISynapse, SynapseOptions, StorageOptions, TokenAmount } from './types.js'
+import { ethers } from 'ethers'
+import type {
+  Synapse as ISynapse,
+  SynapseOptions,
+  StorageOptions,
+  TokenAmount,
+  TokenIdentifier
+} from './types.js'
 import { MockStorageService } from './storage-service.js'
+import {
+  USDFC_ADDRESSES,
+  CHAIN_IDS,
+  ERC20_ABI,
+  PAYMENTS_ADDRESSES,
+  PAYMENTS_ABI
+} from './constants.js'
 
-export class MockSynapse implements ISynapse {
-  private _options: SynapseOptions
-  private _balance: number = 100 // Mock starting balance
-
-  constructor(options: SynapseOptions) {
-    this._options = options
-    console.log('[Mock] Synapse initialized with options:', {
-      withCDN: options.withCDN,
-      rpcAPI: options.rpcAPI,
-      serviceContract: options.serviceContract,
-    })
+export class Synapse implements ISynapse {
+  private readonly _provider: ethers.Provider
+  private readonly _signer: ethers.Signer
+  private readonly _network: 'mainnet' | 'calibration'
+  private readonly _disableNonceManager: boolean
+
+  // Cached contract instances
+  private _usdfcContract: ethers.Contract | null = null
+  private _paymentsContract: ethers.Contract | null = null
+
+  // Static constant for USDFC token identifier
+  static readonly USDFC = 'USDFC' as const
+
+  /**
+   * Create a new Synapse instance with async initialization.
+   * @param options - Configuration options for Synapse
+   * @returns A fully initialized Synapse instance
+   */
+  static async create (options: SynapseOptions): Promise<Synapse> {
+    // Validate options
+    const providedOptions = [options.privateKey, options.provider, options.signer].filter(Boolean).length
+    if (providedOptions !== 1) {
+      throw new Error('Must provide exactly one of: privateKey, provider, or signer')
+    }
+    if (options.privateKey != null && options.rpcURL == null) {
+      throw new Error('rpcURL is required when using privateKey')
+    }
+
+    // Initialize provider and signer
+    let provider: ethers.Provider
+    let signer: ethers.Signer
+
+    if (options.privateKey != null && options.rpcURL != null) {
+      // Create provider from RPC URL
+      if (options.rpcURL.startsWith('ws://') || options.rpcURL.startsWith('wss://')) {
+        provider = new ethers.WebSocketProvider(options.rpcURL)
+      } else {
+        provider = new ethers.JsonRpcProvider(options.rpcURL)
```

Comment:
With GLIF, the free tier is limited to 100 requests per minute. For machines using the SDK, that might not be enough. You can configure GLIF auth keys like so:

```suggestion
        const fetchRequest = new ethers.FetchRequest(options.rpcURL)
        fetchRequest.setHeader('Authorization', `Bearer ${GLIF_TOKEN}`)
        provider = new ethers.JsonRpcProvider(fetchRequest)
```

Reply:
can you do something similar with the websockets rpc provider?

Reply:
ok, I've added an `authorization` option to `SynapseOptions` that will set this up for you if you provide it

although, I think that you should be able to set this up externally with ethers and provide your composite signer/provider instead of giving it your private key and all of this, we're just providing some easy paths here by even accepting a privateKey option

Reply:
I believe you _can't_ do this with websockets. If you want to listen to events, without websockets, pass `{ polling: true }` to the provider options, in which case it will use `eth_getLogs` and not `eth_newFilter`, and that works via https.